WalletWasabi/Wallets/Wallet.cs

```diff
@@ -460,7 +378,11 @@ private async void IndexDownloader_ReorgedAsync(object sender, FilterModel inval
 				using (await HandleFiltersLock.LockAsync())
 				{
 					uint256 invalidBlockHash = invalidFilter.Header.BlockHash;
-					await DeleteBlockAsync(invalidBlockHash);
+					if (BlockProvider is CachedBlockProvider blockCache)
```

Comment:
If we declare `BlockProvider` as a `CachedBlockProvider` in the wallet we will never be able to pass anything different. For example, if I want to pass a fake `IBlockProvider` to unit test the wallet I can do it now or, imagine we decide to use the blocks already downloaded by the bitcoin knots local instance or get the blocks from rpc or any other source, in those cases the `IBlockProvider` is the correct approach. 

I can see that something looks unusual here but I am not sure how to improve it because I can only `Invalidate` something that is cached and only one `IBlockProvider` caches the entries and that is the `CachedBlockProvider`. I will continue thinking about this issue.

Reply:
I dig into this and I see what is happening here. This can be an optimal solution. The other to consider is to add, invalidation to `IBlockProvider` which is implemented with a simple return in `P2pBlockProvider`.
In this way, anyone who inherits from IBlockProvider will know there are situations when a block needs to be deleted.

Reply:
I finally understood what is wrong here: the code that downloads and returns the blocks has more than one responsibility because it also tries to save space in the disk. 

We are deleting a block here because we detected a reorg and we assume that that block will not be part of the winning chain, most of the times that is true but in some cases we will need to download the deleted block again. 

Anyway, the semantic makes clear that the `IBlockProvider` implementations have only one responsibility:  provide blocks. Deleting blocks or saving disk space is not something that the Wallet class has to care about neither. 

If we agree (debatable) that every time we detect a reorg we have to delete the filter and the associated block then the same component should do it, in this case that component could be the `BitcoinStore` class and not the Wallet class. 

-----------

From a design point of view adding a Delete or Invalidate method to the `IBlockProvider` interface would move the smell to somewhere else because those implementations that cannot delete blocks (for example a remote block feching provider) would look like: 

```c#
public void Invalidate(uint256 hash)
{
    // Do nothing
    // Or throw new NotSupportedException("blah blah blah...");
}
```

But this is just anecdotal because the real problem is to understand why we have the bad-looking line that makes clear something is not okay in the design.  

-------------------

Summarizing: BlockProviders have to provide blocks, that's all. For this reason the `CachedBlockProvider.Invalidate` method is an aberration! We should consider to transfer the responsibility over block storage space to the `BitcoinStorage` because it is the one that persists transactions, mempool and indexes, it is also the one that detects the reorgs and removes the invalidated indexes.    

-------------------


As I said in a previous comment I need to think more about this. However I also think this can be done in a different PR.

Reply:
ACK!